Android/app/src/main/java/app/intra/DnsVpnController.java

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2018 Jigsaw Operations LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package app.intra;
+
+import androidx.localbroadcastmanager.content.LocalBroadcastManager;
+import app.intra.util.DnsQueryTracker;
+import app.intra.util.Names;
+
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+
+// Singleton class to maintain state related to VPN Tunnel service.
+public class DnsVpnController {
+
+  private static DnsVpnController dnsVpnServiceState;
+
+  public Object clone() throws CloneNotSupportedException {
+    throw new CloneNotSupportedException();
+  }
+
+  public static synchronized DnsVpnController getInstance() {
```

Comment:
Would it be cleaner to pass the context here instead of passing it in all methods?

Reply:
I tried this, but I got a compiler warning about holding a static reference to `context`.  As I understand it, the problem is that the statically held `context` is actually `MainActivity`, and we want to make sure that there are no long-lived references to `MainActivity`, so that it can be garbage-collected even though the service (and the `DnsVpnController` singleton) remain.